rust/projects/file-io/src/kv.rs

```diff
@@ -187,20 +201,50 @@ impl KvLog {
         if let Some(cmd_pos) = self.index.get(&key) {
             self.reader.seek(SeekFrom::Start(cmd_pos.pos))?;
             let cmd_reader = (&mut self.reader).take(cmd_pos.len);
-            let set_cmd: SetCommand = serde_json::from_reader(cmd_reader)?;
-            Ok(Some(set_cmd.value))
+            if let Command::Set { value, .. } = serde_json::from_reader(cmd_reader)? {
+                Ok(Some(value))
+            } else {
+                Err(KvsError::UnexpectedCommandType)
+            }
         } else {
             Ok(None)
         }
     }
 
+    fn remove(&mut self, key: String) -> Result<()> {
+        assert!(self.loaded);
+
+        if self.index.contains_key(&key) {
+            let cmd = Command::remove(key);
+            serde_json::to_writer(&mut self.writer, &cmd)?;
+            self.writer.flush()?;
+            if let Command::Remove { key } = cmd {
+                let old_cmd = self.index.remove(&key).expect("key not found");
+                self.uncompacted += old_cmd.len;
+            }
+            Ok(())
+        } else {
+            Err(KvsError::KeyNotFound)
+        }
+    }
+
     fn load(&mut self) -> Result<()> {
         let mut pos = self.reader.seek(SeekFrom::Start(0))?;
-        let mut stream = Deserializer::from_reader(&mut self.reader).into_iter::<SetCommand>();
-        while let Some(set_cmd) = stream.next() {
+        let mut stream = Deserializer::from_reader(&mut self.reader).into_iter::<Command>();
+        while let Some(cmd) = stream.next() {
             let new_pos = stream.byte_offset() as u64;
-            if let Some(old_cmd) = self.index.insert(set_cmd?.key, (pos..new_pos).into()) {
-                self.uncompacted += old_cmd.len;
+            match cmd? {
+                Command::Set { key, .. } => {
+                    if let Some(old_cmd) = self.index.insert(key, (pos..new_pos).into()) {
+                        self.uncompacted += old_cmd.len;
+                    }
+                }
+                Command::Remove { key } => {
+                    if let Some(old_cmd) = self.index.remove(&key) {
+                        self.uncompacted += old_cmd.len;
+                    }
+                    self.uncompacted += new_pos - pos;
```

Comment:
It took me a surprising amount of staring at it to understand the logic of this line. I think more comments about the `uncompacted` field is warranted. The field could be annotated with something like "The number of bytes representing 'stale' commands that could be deleted during a compaction", and this line with "since the key no longer exists, the 'remove' command itself can be deleted in the next compaction".